[OID4VCI] Limiting Client capabilities to OpenID Connect or OpenID4VCI

forkimenjeckayang opened the pull request:
This PR aims to limit client capabilities to OpenID Connect or OpenID4VCI . In order to acheive this, we have a new section in the front end under the advanced tab of the client settings configurable for OpenID for Verifiable Credentials. This setting is limited or will be available only to clients operating under the OpenID Connect or OpenID4VCI protocol.

![Screenshot from 2025-04-24 13-32-49](https://github.com/user-attachments/assets/5f7bc761-e211-48b2-a15e-7204f28003d7)

Once toggled on and saved, we have an attribute assigned to that client oid4vci.enabled and set to true
This signifies or enables the client to be used to request verfiable credentials from Keycloak's credential endpoint. After client authentication this attribute is checked and if not available for the client, it cannot be used for requestig credentials.

Closes https://github.com/keycloak/keycloak/issues/32967

ssilvert reviewed (changes requested):
I haven't reviewed all the UI stuff yet, but I noticed that you are adding translations for every language.  Did you use Google translate or something like that?

What we typically do is to update English only and leave updates of the rest to native speakers.  So there is no need to add key/value pairs for other languages unless you are proficient in that language.

forkimenjeckayang commented:
> I haven't reviewed all the UI stuff yet, but I noticed that you are adding translations for every language. Did you use Google translate or something like that?
> 
> What we typically do is to update English only and leave updates of the rest to native speakers. So there is no need to add key/value pairs for other languages unless you are proficient in that language.

Thanks for pointing that out @ssilvert , and yes I did use google translate. In that case i will leave just the english version and clear out the others.

edewit commented on js/apps/admin-ui/src/clients/advanced/OpenIdVerifiableCredentials.tsx:
```diff
@@ -0,0 +1,92 @@
+        fineGrainedAccess={hasConfigureAccess}
+        isHorizontal
+      >
+        <FormGroup
+          label={t("oid4vciEnabled")}
+          fieldId="oid4vciEnabled"
+          hasNoPaddingTop
+          labelIcon={
+            <HelpItem
+              helpText={t("oid4vciEnabledHelp")}
+              fieldLabelId="oid4vciEnabled"
+            />
+          }
+        >
+          <Controller
+            name={convertAttributeNameToForm<FormFields>(
+              "attributes.oid4vci.enabled",
+            )}
+            defaultValue={client.attributes?.["oid4vci.enabled"] || "false"}
+            control={control}
+            render={({ field }) => (
+              <Switch
+                id="oid4vciEnabled-switch"
+                label={t("on")}
+                labelOff={t("off")}
+                isChecked={field.value === "true"}
+                onChange={(_event, value) => field.onChange(value.toString())}
+                aria-label={t("oid4vciEnabled")}
+              />
+            )}
+          />
```
You can use the `DefaultSwitchControl` instead:
```suggestion
          <DefaultSwitchControl
              name={convertAttributeNameToForm<FormFields>(
                 "attributes.oid4vci.enabled",
              )}
              label={t("oid4vciEnabled")}
              labelIcon={t("oid4vciEnabledHelp")}
              stringify
            />
```

forkimenjeckayang committed:
Add OpenID for Verifiable Credentials support in client settings

Closes #32967

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

forkimenjeckayang commented:
Thanks for your review @edewit 
I addressed the changes you pointed out
Could you please check again?

Captain-P-Goldfish commented:
> I haven't reviewed all the UI stuff yet, but I noticed that you are adding translations for every language. Did you use Google translate or something like that?
> 
> What we typically do is to update English only and leave updates of the rest to native speakers. So there is no need to add key/value pairs for other languages unless you are proficient in that language.

From my latest experiences, ChatGPT is doing a great job when it comes to translations. Makes probably still some mistakes now and then but I was pretty thrilled by japanese to german translations and other languages.

forkimenjeckayang committed:
update: removed tests for Verifiable Credentials support in client settings for protocol oid4vc

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

forkimenjeckayang commented on js/apps/admin-ui/test/clients/advanced.spec.ts:
```diff
@@ -136,3 +140,74 @@ test.describe("Client Offline Session Max", () => {
+      realm: realmName,
+      protocol: "oid4vc",
+    });
+  });
```
> Could you remove this test again? This protocol-type is no longer supported for clients. The protocol-type `oid4vc` is used on ClientScopes instead and clients will still have protocol-type `openid-connect`

Thanks for pointing that out @Captain-P-Goldfish 
I have done the necessary changes in the tetst file 
Please could you check again?

forkimenjeckayang committed:
update: removed unused imports in OpenIdVerifiableCredentials.tsx

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

mposolda commented on js/apps/admin-ui/src/clients/AdvancedTab.tsx:
```diff
@@ -199,6 +201,17 @@ export const AdvancedTab = ({ save, client }: AdvancedProps) => {
           },
+          {
+            title: t("openIdVerifiableCredentials"),
+            isHidden: protocol !== openIdConnect && protocol !== oid4vc,
```
Besides this check, it seems this element should be shown just if `oid4vc_vci` feature itself is enabled?

forkimenjeckayang replied:
Exactly, that's how it should behave

mposolda reviewed (commented):
Added one review check inline. Could you check it?

Besides that, I am not sure about changes in files `messages_nl` , `messages_ro` and `messages_sl` as there is only single line deletion in those files. Was it by accident? Maybe better to avoid the change in these files entirely?

forkimenjeckayang commented:
> Added one review check inline. Could you check it?
> 
> Besides that, I am not sure about changes in files `messages_nl` , `messages_ro` and `messages_sl` as there is only single line deletion in those files. Was it by accident? Maybe better to avoid the change in these files entirely?

About that , i included translations for those languages and decided to remove them based on what @ssilvert said. I'll revert the changes i made there correctly

forkimenjeckayang committed:
Merge branch 'main' into issue-32967

forkimenjeckayang committed:
update: included extra check for oid4vc feature enabled and refactored tests

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

forkimenjeckayang commented:
Hello @mposolda 
I've addressed the changes you suggested.
We now have the openIdVerifiableCredentials section displayed when the feature `oid4vc_vci`  is enabled. If not enabled then it's not displayed under the advanced tab section. Refactored tests too to fit that purpose.
Please have another look and let me know
Thank you

forkimenjeckayang committed:
fix: resolved merge conflicts

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

forkimenjeckayang committed:
Merge branch 'main' into issue-32967

forkimenjeckayang committed:
Merge branch 'main' into issue-32967

francis-pouatcha committed:
Merge branch 'main' into issue-32967

forkimenjeckayang committed:
Merge branch 'main' into issue-32967

Signed-off-by: forkimenjeckayang <104195313+forkimenjeckayang@users.noreply.github.com>

forkimenjeckayang committed:
update: resolved conflicts in messages_ru.properties

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

forkimenjeckayang committed:
update:message properties files

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

forkimenjeckayang committed:
update:message properties files

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

mposolda reviewed (dismissed):
@forkimenjeckayang Thanks for the updates!
@Captain-P-Goldfish @francis-pouatcha Thanks for your review!

I am approving the PR. It will need also another review from our UI team. I hope it can be merged afterwards.

jonkoops reviewed (changes requested):
Just a couple of comments/questions.

forkimenjeckayang committed:
refactor: clarify protocol constants, section visibility logic, and move help text for OID4VC section

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

forkimenjeckayang committed:
fix: resolved merge conflicts

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

forkimenjeckayang committed:
Merge branch 'main' into issue-32967

forkimenjeckayang committed:
Merge branch 'main' into issue-32967

forkimenjeckayang committed:
fix: resolved merged conflicts and align with new main implementation

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

forkimenjeckayang committed:
fix: resolved merged

Signed-off-by: forkimenjeckayang <forkimenjeckayang@gmail.com>

jonkoops reviewed (approved):
LGTM

mposolda reviewed (approved):
@forkimenjeckayang Thanks!
@jonkoops Thanks for the additional review!
